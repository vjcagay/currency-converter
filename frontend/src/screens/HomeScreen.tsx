import React, { useState } from "react";
import CurrencyPicker from "../components/currency-picker/CurrencyPicker";

const HomeScreen = () => {
  const [from, setFrom] = useState("USD");
  const [to, setTo] = useState("JPY");

  return (
    <>
      <CurrencyPicker from={from} to={to} onFromChange={setFrom} onToChange={setTo} />
    </>
  );
};

export default HomeScreen;
