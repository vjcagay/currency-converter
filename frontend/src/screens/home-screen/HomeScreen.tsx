import React, { useEffect, useState } from "react";
import CurrencyPicker from "../../components/currency-picker/CurrencyPicker";
import DatePicker from "../../components/date-picker/DatePicker";
import LineGraph from "../../components/line-graph/LineGraph";
import Summary from "../../components/summary/Summary";
import styles from "./styles.module.css";

type FetchResponseJSON = {
  start: string;
  end: string;
  from: string;
  to: string;
  gainLossPercentage: number;
  data: { date: string; rate: string }[];
};

const HomeScreen = () => {
  const [from, setFrom] = useState("USD");
  const [to, setTo] = useState("JPY");

  const [start, setStart] = useState("");
  const [end, setEnd] = useState("");

  const [dates, setDates] = useState<string[]>([]);
  const [rates, setRates] = useState<number[]>([]);

  const [presetRange, setPresetRange] = useState("1W");
  const [gainLoss, setGainLoss] = useState(0);

  useEffect(() => {
    if (from && to && start && end) {
      fetch(`http://localhost:8081/history?from=${from}&to=${to}&start=${start}&end=${end}`)
        .then((results) => {
          return results.json();
        })
        .then((json: FetchResponseJSON) => {
          const { dates, rates } = json.data.reduce(
            (acc, item) => {
              acc.dates.push(item.date);
              acc.rates.push(item.rate);

              return acc;
            },
            { dates: [], rates: [] },
          );

          setDates(dates);
          setRates(rates);
          setGainLoss(json.gainLossPercentage);
        });
    }
  }, [from, to, start, end]);

  return (
    <>
      <CurrencyPicker
        className={styles.currencyPicker}
        from={from}
        to={to}
        onChange={(value) => {
          setFrom(value.from);
          setTo(value.to);
        }}
      />
      <Summary className={styles.summary} from={from} to={to} gainLossPercentage={gainLoss} rangeString={presetRange} />
      <DatePicker
        className={styles.datePicker}
        start={start}
        end={end}
        onChange={(value, rangeName) => {
          setStart(value.start);
          setEnd(value.end);
          setPresetRange(rangeName);
        }}
      />
      <LineGraph
        labels={dates}
        datasets={[
          {
            label: "Exchange Rate",
            data: rates,
            borderColor: "#0171eb",
            backgroundColor: "#0171eb",
            tension: 0.2,
          },
        ]}
      />
    </>
  );
};

export default HomeScreen;
