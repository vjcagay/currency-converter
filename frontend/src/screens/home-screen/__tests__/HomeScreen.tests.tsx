import { render, waitFor } from "@testing-library/react";
import React from "react";
import HomeScreen from "../HomeScreen";

(global as any).fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        start: "2023-09-26",
        end: "2023-09-30",
        from: "USD",
        to: "JPY",
        gainLossPercentage: -0.032,
        data: [
          {
            date: "2023-09-30",
            rate: 149.235416,
          },
          {
            date: "2023-09-29",
            rate: 149.235416,
          },
          {
            date: "2023-09-28",
            rate: 149.283613,
          },
        ],
      }),
  }),
);

// Used by Chart.js
global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

describe("<HomeScreen />", () => {
  it("should render correctly with a gain percentage", async () => {
    const { asFragment } = render(<HomeScreen />);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
