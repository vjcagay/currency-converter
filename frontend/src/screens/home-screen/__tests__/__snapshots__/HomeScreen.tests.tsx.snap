// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> should render correctly with a gain percentage 1`] = `
<DocumentFragment>
  <div
    class="wrapper currencyPicker"
  >
    <label
      class="label"
    >
      <span
        class="labelText"
      >
        From
      </span>
      <select
        class="select"
        data-testid="currency-picker-from"
      >
        <option
          value="USD"
        >
          ðŸ‡ºðŸ‡¸ USD - US Dollar
        </option>
        <option
          value="SGD"
        >
          ðŸ‡¸ðŸ‡¬ SGD - Singapore Dollar
        </option>
        <option
          value="JPY"
        >
          ðŸ‡¯ðŸ‡µ JPY - Japanese Yen
        </option>
        <option
          value="EUR"
        >
          ðŸ‡ªðŸ‡º EUR - Euro
        </option>
      </select>
    </label>
    <button
      class="button"
      data-testid="currency-picker-reverse"
    />
    <label
      class="label"
    >
      <span
        class="labelText"
      >
        To
      </span>
      <select
        class="select"
        data-testid="currency-picker-to"
      >
        <option
          value="USD"
        >
          ðŸ‡ºðŸ‡¸ USD - US Dollar
        </option>
        <option
          value="SGD"
        >
          ðŸ‡¸ðŸ‡¬ SGD - Singapore Dollar
        </option>
        <option
          value="JPY"
        >
          ðŸ‡¯ðŸ‡µ JPY - Japanese Yen
        </option>
        <option
          value="EUR"
        >
          ðŸ‡ªðŸ‡º EUR - Euro
        </option>
      </select>
    </label>
  </div>
  <div
    class="wrapper summary"
  >
    <b
      class="titleText"
    >
      USD to JPY Chart
    </b>
    <wbr />
    Â 
    <span
      class="gainLossText gain"
    >
      +0%
    </span>
    Â 
    <span
      class="rangeText"
    >
      (1W)
    </span>
    <br />
    <small>
      US Dollar to Japanese Yen
    </small>
  </div>
  <div
    class="wrapper datePicker"
  >
    <fieldset
      class="fieldset presetRangeFieldset"
    >
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-1W"
      >
        <input
          checked=""
          type="radio"
        />
        <span>
          1W
        </span>
      </label>
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-1M"
      >
        <input
          type="radio"
        />
        <span>
          1M
        </span>
      </label>
      <label
        class="presetRangeLabel"
        data-testid="date-picker-range-6M"
      >
        <input
          type="radio"
        />
        <span>
          6M
        </span>
      </label>
    </fieldset>
    <fieldset
      class="fieldset customRangeFieldset"
    >
      <button
        class="customRangeFieldsetButton"
        data-testid="date-picker-open-calendar"
      >
        <img
          alt="calendar icon"
          src="test-file-stub"
        />
      </button>
    </fieldset>
  </div>
  <canvas
    class="wrapper "
    height="0"
    style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
    width="0"
  />
</DocumentFragment>
`;
